### DISCUSSION TITLE
Unclear interface needs improvement

### DISCUSSION THREAD

**Discussion by nav9** [memtest86plus/memtest86plus#323] [Enhancement]
Firstly, thank you for this wonderful software. However, there are some things about the interface that need changes, which I hope you'd consider and implement.   
This is for Memtest86+ v6.10.   
* **How many tests and how long:** When you run tests, it helps to know exactly how many tests you will be running, and how long it'd take to complete all of them. 
* **Beep:** It helps to provide a beep sound when the program passes or fails the tests. Internal speaker. Yes, some of us still have it.
* **Scroll:** When the errors are listed on the screen, there's no option to scroll up or down to view all of them.
* **Suggestions:** Memtest86 showed that there were errors when testing my RAM, and I thought I'd have to replace my RAM. But thankfully, after a BIOS reset to defaults, the tests passed. I later realized that it was due to XMP being switched off, that the errors didn't happen. I know you've mentioned it on your [site](https://www.memtest.org/) as "_If your computer is overclocked, you should try again at default settings first._ ". However, please remember that not everyone knows to visit your site for this info. It helps to provide this info within the memtest interface itself. So when you show the "Fail" message, you also need to show a message with suggestions about what a User could try to fix the errors. 
* **Dark screen:** Rather than have the blue background, couldn't you provide a black or grey background? It'd consume lesser power and be easy on the eyes too.
* **Stopping on single error:** From what I understand, even a single error detected, is enough to proclaim the RAM faulty. So couldn't the program stop testing when it finds even one error, and beep to let the User know. It'd help further to show a message informing the User that even a single error means there's a serious problem, and then show them suggestions of what they could do to fix it.
* Explanations: It'd help to provide a hotkey and dialog box to show explanations about the various strings shown on screen. Eg: what is pCPU, what is CLK.
* **Which RAM is problematic:** I have two RAM sticks, and when errors were detected, the interface didn't inform me about which RAM the problems were detected in. 
* **When to stop testing:** As of now, when the pass or fail message is displayed, I see in the top right side of the screen, some tests are still running. This is confusing. Does it mean that only one set of tests passed and more need to be run? As I mentioned earlier, the interface provides very poor information on how many tests are going to be run and how long it'd take to complete all of them. This really needs to be improved.  
  
If you do a google search, you'll find other users also asking about various aspects of memtest, because it's so confusing to understand. Please re-think and re-design the user interface. After improving it, you can do a "hallway testing", by inviting random people to try out the interface and getting their opinion on what they understand and what they find confusing about the interface.

**Comment by debrouxl** [comment]
Thanks for your feedback.
As a preliminary note: UI-wise, there's only so much that can be done on a 80x25 character screen.

* **How many tests and how long**:
    *  the number of tests isn't part of the main screen, but it's accessible through the configuration menu. One of the tests is disabled by default on x86-64; displaying test number + number of enabled tests + number of possible tests might bring more confusion than value ?
    * obtaining a reliable estimation for the ETA is a *hard* problem, due to the effect of caches, NUMA, etc. I'd argue that it's better not to do it than to either use simplistic code which would provide Windows explorer file operation ETA level reliability, or make parts of the code base significantly more complex in order to try and provide more reasonable ETA figures :)
* **Beep**: unfortunately, that's also a harder problem than many people think, see my summary in one comment from [#172](https://github.com/memtest86plus/memtest86plus/issues/172#issuecomment-1477427005) .
* **Scroll**: it could be argued that either there are few errors, in which case the current shallow memoization is enough badram-like mechanisms may provide useful workarounds, or there are many errors, and the faulty RAM sticks really should be considered untrusted for anything and be replaced. With that dichotomy, a mechanism to store many errors and scroll through the list wouldn't seem to be high priority, would it ?
* **Suggestions**: point taken, though I'd say that if a motherboard + memory stick combo passes tests at JEDEC speeds but doesn't at XMP speeds, at least one of the sides of the combo is flaky. Also, 80x25 screen.
* **Dark screen**: technically, dark or high contrast modes aren't hard to implement - that is, for compile-time selection between the modes. Run-time selection is more work;
* **Stopping on single error**: the big PASS or FAIL messages, nowadays enabled by default, convey a hopefully readable message about an error being detected. On principle, I'm not against adding several words of text (beyond the ones below the big PASS or FAIL messages), but... 80x25 screen ;)
* **Explanations**: a glossary page... eh, why not, as long as the information to access it does not clutter the cramped 80x25 screen;
* **Which RAM is problematic**: that's also a hard problem, in that there are many models of memory controllers, not all of which provide such functionality. Implementing support for every single model of memory controller takes time (an issue for a volunteer project like Memtest86+, or most open-source projects anyway, are) and requires having extensive physical access to motherboards + CPUs which contain the given memory controller model, as well as RAM sticks which contain reliably reproducible errors - those are surprisingly hard to come by for maintainers.
* **When to stop testing**: basically, after the first pass, though the program runs until the user reboots the computer. Subsequent passes will repeat the tests with nearly the same parameters (the bit fade test delay increases a bit).

**Comment by nav9** [comment]
I'm grateful for the detailed reply. I'm from the GW-BASIC and Q-BASIC era, so grew up with [scrollbars](https://upload.wikimedia.org/wikipedia/en/0/01/QBasic_Opening_Screen.png) in an ASCII screen and accurate ETA estimations in the old Norton Disk Doctor. I understand it'd be hard to bring in such changes in Memtest since it wasn't planned beforehand.  
So I'll mention the essence of the confusion which would be easier to fix:   
1. In the top right section, you display "Pass", "Test" and test number. As a user, I basically want to know how many of these tests are going to be run and whether each pass runs the same bunch of tests again, and what is the ideal number of passes I need to run to determine if my RAM is working well (because at 3200MHz, the tests succeed during the first pass and failed during the second pass).  Rather than display a huge "PASS" or "FAIL" message, it'd be more helpful to show a list like: Pass1: Success. Pass2: Failed. Pass3: Success....and so on.  
Idea: You display "Pass" and "Test" as percentages, which is good, but space can be saved if instead of a progress bar, only the percentage is shown. Perhaps like "Pass: 86/100", indicating 86%.
2. In the area where you show "Time" and "Status", there's also another "Pass". This confuses the user, because the user would wonder if this pass has anything to do with the above pass, and how many more passes are going to be run.  
3. The "Suggestions" point in my initial post, about giving the user pointers on what they could do. Perhaps providing a hyperlink to a page containing useful info.    
   
In this reply, all I'm mentioning is where the core confusion is, so that (if you deem it appropriate) it'd help you come up with ideas on how to make the interface more intuitive.  
  
_ps: I've [been in touch](https://forums.linuxmint.com/viewtopic.php?t=398408) with the Mint OS team and requested them to add Memtest as a default grub option. I'm also in the process of discussing the RAM issue with the MSI team, and they say that  while your explanation of one side of the RAM being flaky may be possible, there are also other potential factors to consider, like: 1) The mainboard model (PCB layer count, PCB trace optimization, RAM slot topology and slot count, component selection, RAM VRM etc.), 2) The mainboard's BIOS optimizations and the BIOS settings, 3) The CPU's integrated memory controller (IMC), quality depends on the CPU generation as well as on the individual CPU (silicon lottery), 4) The properties of the RAM modules._
